Error when running PyTorch example

I run pytorch/train/mnist/main.py on YARN Almaren cluster and get the following error:
```
[Stage 1:>                                                          (0 + 2) / 2]2020-08-10 18:43:28 ERROR TaskSetManager:70 - Task 1 in stage 1.0 failed 1 times; aborting job
Traceback (most recent call last):
  File "main.py", line 124, in <module>
    main()
  File "main.py", line 113, in main
    train_featureset = FeatureSet.pytorch_dataloader(train_loader)
  File "/opt/work/client/anaconda3/envs/standalone/lib/python3.6/site-packages/zoo/feature/common.py", line 383, in pytorch_dataloader
    jvalue = callZooFunc(bigdl_type, "createFeatureSetFromPyTorch", bys)
  File "/opt/work/client/anaconda3/envs/standalone/lib/python3.6/site-packages/zoo/common/utils.py", line 133, in callZooFunc
    raise e
  File "/opt/work/client/anaconda3/envs/standalone/lib/python3.6/site-packages/zoo/common/utils.py", line 127, in callZooFunc
    java_result = api(*args)
  File "/opt/work/client/anaconda3/envs/standalone/lib/python3.6/site-packages/py4j/java_gateway.py", line 1257, in __call__
    answer, self.gateway_client, self.target_id, self.name)
  File "/opt/work/client/anaconda3/envs/standalone/lib/python3.6/site-packages/py4j/protocol.py", line 328, in get_return_value
    format(target_id, ".", name), value)
py4j.protocol.Py4JJavaError: An error occurred while calling o67.createFeatureSetFromPyTorch.
: org.apache.spark.SparkException: Job aborted due to stage failure: Task 1 in stage 1.0 failed 1 times, most recent failure: Lost task 1.0 in stage 1.0 (TID 3, Almaren-Node-058, executor 2): jep.JepException: jep.JepException: <class 'zipimport.ZipImportError'>: can't decompress data; zlib not available
        at com.intel.analytics.zoo.common.PythonInterpreter$.threadExecute(PythonInterpreter.scala:87)
        at com.intel.analytics.zoo.common.PythonInterpreter$.exec(PythonInterpreter.scala:96)
        at com.intel.analytics.zoo.feature.PythonFeatureSet$$anonfun$loadPythonSet$1.apply(FeatureSet.scala:358)
        at com.intel.analytics.zoo.feature.PythonFeatureSet$$anonfun$loadPythonSet$1.apply(FeatureSet.scala:354)
        at org.apache.spark.rdd.RDD$$anonfun$mapPartitions$1$$anonfun$apply$23.apply(RDD.scala:801)
        at org.apache.spark.rdd.RDD$$anonfun$mapPartitions$1$$anonfun$apply$23.apply(RDD.scala:801)
        at org.apache.spark.rdd.MapPartitionsRDD.compute(MapPartitionsRDD.scala:52)
        at org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:324)
        at org.apache.spark.rdd.RDD.iterator(RDD.scala:288)
        at org.apache.spark.scheduler.ResultTask.runTask(ResultTask.scala:90)
        at org.apache.spark.scheduler.Task.run(Task.scala:121)
        at org.apache.spark.executor.Executor$TaskRunner$$anonfun$10.apply(Executor.scala:408)
        at org.apache.spark.util.Utils$.tryWithSafeFinally(Utils.scala:1360)
        at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:414)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
        at java.lang.Thread.run(Thread.java:745)
Caused by: jep.JepException: <class 'zipimport.ZipImportError'>: can't decompress data; zlib not available
        at <string>.<module>(<string>:2)
        at jep.Jep.exec(Native Method)
        at jep.Jep.exec(Jep.java:478)
        at com.intel.analytics.zoo.common.PythonInterpreter$$anonfun$1.apply$mcV$sp(PythonInterpreter.scala:94)
        at com.intel.analytics.zoo.common.PythonInterpreter$$anonfun$1.apply(PythonInterpreter.scala:93)
        at com.intel.analytics.zoo.common.PythonInterpreter$$anonfun$1.apply(PythonInterpreter.scala:93)
        at scala.concurrent.impl.Future$PromiseCompletingRunnable.liftedTree1$1(Future.scala:24)
        at scala.concurrent.impl.Future$PromiseCompletingRunnable.run(Future.scala:24)
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
        at java.util.concurrent.FutureTask.run(FutureTask.java:266)
        ... 3 more

Driver stacktrace:
        at org.apache.spark.scheduler.DAGScheduler.org$apache$spark$scheduler$DAGScheduler$$failJobAndIndependentStages(DAGScheduler.scala:1889)
        at org.apache.spark.scheduler.DAGScheduler$$anonfun$abortStage$1.apply(DAGScheduler.scala:1877)
        at org.apache.spark.scheduler.DAGScheduler$$anonfun$abortStage$1.apply(DAGScheduler.scala:1876)
        at scala.collection.mutable.ResizableArray$class.foreach(ResizableArray.scala:59)
        at scala.collection.mutable.ArrayBuffer.foreach(ArrayBuffer.scala:48)
        at org.apache.spark.scheduler.DAGScheduler.abortStage(DAGScheduler.scala:1876)
        at org.apache.spark.scheduler.DAGScheduler$$anonfun$handleTaskSetFailed$1.apply(DAGScheduler.scala:926)
        at org.apache.spark.scheduler.DAGScheduler$$anonfun$handleTaskSetFailed$1.apply(DAGScheduler.scala:926)
        at scala.Option.foreach(Option.scala:257)
        at org.apache.spark.scheduler.DAGScheduler.handleTaskSetFailed(DAGScheduler.scala:926)
        at org.apache.spark.scheduler.DAGSchedulerEventProcessLoop.doOnReceive(DAGScheduler.scala:2110)
        at org.apache.spark.scheduler.DAGSchedulerEventProcessLoop.onReceive(DAGScheduler.scala:2059)
        at org.apache.spark.scheduler.DAGSchedulerEventProcessLoop.onReceive(DAGScheduler.scala:2048)
        at org.apache.spark.util.EventLoop$$anon$1.run(EventLoop.scala:49)
        at org.apache.spark.scheduler.DAGScheduler.runJob(DAGScheduler.scala:737)
        at org.apache.spark.SparkContext.runJob(SparkContext.scala:2061)
        at org.apache.spark.SparkContext.runJob(SparkContext.scala:2082)
        at org.apache.spark.SparkContext.runJob(SparkContext.scala:2101)
        at org.apache.spark.SparkContext.runJob(SparkContext.scala:2126)
        at org.apache.spark.rdd.RDD.count(RDD.scala:1168)
        at com.intel.analytics.zoo.feature.PythonFeatureSet$.loadPythonSet(FeatureSet.scala:366)
        at com.intel.analytics.zoo.feature.PythonFeatureSet.<init>(FeatureSet.scala:428)
        at com.intel.analytics.zoo.feature.FeatureSet$.python(FeatureSet.scala:648)
        at com.intel.analytics.zoo.feature.python.PythonFeatureSet.createFeatureSetFromPyTorch(PythonFeatureSet.scala:189)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:497)
        at py4j.reflection.MethodInvoker.invoke(MethodInvoker.java:244)
        at py4j.reflection.ReflectionEngine.invoke(ReflectionEngine.java:357)
        at py4j.Gateway.invoke(Gateway.java:282)
        at py4j.commands.AbstractCommand.invokeMethod(AbstractCommand.java:132)
        at py4j.commands.CallCommand.execute(CallCommand.java:79)
        at py4j.GatewayConnection.run(GatewayConnection.java:238)
        at java.lang.Thread.run(Thread.java:745)
Caused by: jep.JepException: jep.JepException: <class 'zipimport.ZipImportError'>: can't decompress data; zlib not available
        at com.intel.analytics.zoo.common.PythonInterpreter$.threadExecute(PythonInterpreter.scala:87)
        at com.intel.analytics.zoo.common.PythonInterpreter$.exec(PythonInterpreter.scala:96)
        at com.intel.analytics.zoo.feature.PythonFeatureSet$$anonfun$loadPythonSet$1.apply(FeatureSet.scala:358)
        at com.intel.analytics.zoo.feature.PythonFeatureSet$$anonfun$loadPythonSet$1.apply(FeatureSet.scala:354)
        at org.apache.spark.rdd.RDD$$anonfun$mapPartitions$1$$anonfun$apply$23.apply(RDD.scala:801)
        at org.apache.spark.rdd.RDD$$anonfun$mapPartitions$1$$anonfun$apply$23.apply(RDD.scala:801)
        at org.apache.spark.rdd.MapPartitionsRDD.compute(MapPartitionsRDD.scala:52)
        at org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:324)
        at org.apache.spark.rdd.RDD.iterator(RDD.scala:288)
        at org.apache.spark.scheduler.ResultTask.runTask(ResultTask.scala:90)
        at org.apache.spark.scheduler.Task.run(Task.scala:121)
        at org.apache.spark.executor.Executor$TaskRunner$$anonfun$10.apply(Executor.scala:408)
        at org.apache.spark.util.Utils$.tryWithSafeFinally(Utils.scala:1360)
        at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:414)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
        ... 1 more
Caused by: jep.JepException: <class 'zipimport.ZipImportError'>: can't decompress data; zlib not available
        at <string>.<module>(<string>:2)
        at jep.Jep.exec(Native Method)
        at jep.Jep.exec(Jep.java:478)
        at com.intel.analytics.zoo.common.PythonInterpreter$$anonfun$1.apply$mcV$sp(PythonInterpreter.scala:94)
        at com.intel.analytics.zoo.common.PythonInterpreter$$anonfun$1.apply(PythonInterpreter.scala:93)
        at com.intel.analytics.zoo.common.PythonInterpreter$$anonfun$1.apply(PythonInterpreter.scala:93)
        at scala.concurrent.impl.Future$PromiseCompletingRunnable.liftedTree1$1(Future.scala:24)
        at scala.concurrent.impl.Future$PromiseCompletingRunnable.run(Future.scala:24)
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
        at java.util.concurrent.FutureTask.run(FutureTask.java:266)
        ... 3 more
```
Am I doing anything wrong or it is related to executor no zlib?

When I run in local, got the following error:
```
Caused by: jep.JepException: <class 'ImportError'>: /usr/lib/x86_64-linux-gnu/libstdc++.so.6: version `GLIBCXX_3.4.21' not found (required by /opt/work/client/anaconda3/envs/orca-kai/lib/python3.7/site-packages/torch/lib/libshm.so)
        at /opt/work/client/anaconda3/envs/orca-kai/lib/python3.7/site-packages/torch/__init__.<module>(__init__.py:136)
        at <string>.<module>(<string>:6)
        at jep.Jep.exec(Native Method)
        at jep.Jep.exec(Jep.java:478)
        at com.intel.analytics.zoo.common.PythonInterpreter$$anonfun$1.apply$mcV$sp(PythonInterpreter.scala:94)
        at com.intel.analytics.zoo.common.PythonInterpreter$$anonfun$1.apply(PythonInterpreter.scala:93)
        at com.intel.analytics.zoo.common.PythonInterpreter$$anonfun$1.apply(PythonInterpreter.scala:93)
        at scala.concurrent.impl.Future$PromiseCompletingRunnable.liftedTree1$1(Future.scala:24)
        at scala.concurrent.impl.Future$PromiseCompletingRunnable.run(Future.scala:24)
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
        at java.util.concurrent.FutureTask.run(FutureTask.java:266)
        ... 3 more
```

> When I run in local, got the following error:
> 
> ```
> Caused by: jep.JepException: <class 'ImportError'>: /usr/lib/x86_64-linux-gnu/libstdc++.so.6: version `GLIBCXX_3.4.21' not found (required by /opt/work/client/anaconda3/envs/orca-kai/lib/python3.7/site-packages/torch/lib/libshm.so)
>         at /opt/work/client/anaconda3/envs/orca-kai/lib/python3.7/site-packages/torch/__init__.<module>(__init__.py:136)
>         at <string>.<module>(<string>:6)
>         at jep.Jep.exec(Native Method)
>         at jep.Jep.exec(Jep.java:478)
>         at com.intel.analytics.zoo.common.PythonInterpreter$$anonfun$1.apply$mcV$sp(PythonInterpreter.scala:94)
>         at com.intel.analytics.zoo.common.PythonInterpreter$$anonfun$1.apply(PythonInterpreter.scala:93)
>         at com.intel.analytics.zoo.common.PythonInterpreter$$anonfun$1.apply(PythonInterpreter.scala:93)
>         at scala.concurrent.impl.Future$PromiseCompletingRunnable.liftedTree1$1(Future.scala:24)
>         at scala.concurrent.impl.Future$PromiseCompletingRunnable.run(Future.scala:24)
>         at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
>         at java.util.concurrent.FutureTask.run(FutureTask.java:266)
>         ... 3 more
> ```

Seems something wrong in your torch.

--- a/pyzoo/zoo/util/spark.py
+++ b/pyzoo/zoo/util/spark.py
@@ -101,7 +101,7 @@ class SparkRunner:
                                          executor_cores, executor_memory,
                                          extra_executor_memory_for_ray)
             py_version = ".".join(platform.python_version().split(".")[0:2])
-            preload_so = self.PYTHON_ENV + "/lib/libpython" + py_version + "m.so"
+            preload_so = executor_python_env + "/lib/libpython" + py_version + "m.so"
             conf.update({"spark.scheduler.minRegisteredResourcesRatio": "1.0",
                          "spark.executorEnv.PYTHONHOME": executor_python_env,
                          "spark.executorEnv.LD_PRELOAD": preload_so})

--- a/pyzoo/zoo/util/spark.py
+++ b/pyzoo/zoo/util/spark.py
@@ -15,6 +15,7 @@
 #
 
 import os
+import platform
 
 from pyspark import SparkContext
 from zoo import init_nncontext, init_spark_conf
@@ -99,8 +100,11 @@ class SparkRunner:
             conf = enrich_conf_for_spark(conf, driver_cores, driver_memory, num_executors,
                                          executor_cores, executor_memory,
                                          extra_executor_memory_for_ray)
+            py_version = ".".join(platform.python_version().split(".")[0:2])
+            preload_so = self.PYTHON_ENV + "/lib/libpython" + py_version + "m.so"
             conf.update({"spark.scheduler.minRegisteredResourcesRatio": "1.0",
-                         "spark.executorEnv.PYTHONHOME": executor_python_env})
+                         "spark.executorEnv.PYTHONHOME": executor_python_env,
+                         "spark.executorEnv.LD_PRELOAD": preload_so})
             if spark_yarn_archive:
                 conf["spark.yarn.archive"] = spark_yarn_archive
             zoo_bigdl_path_on_executor = ":".join(

